cluster-autoscaler/core/scaledown/actuation/actuator.go

```diff
@@ -251,41 +246,50 @@ func (a *Actuator) taintNode(node *apiv1.Node) error {
 	return nil
 }
 
-// deleteNode performs the deletion of the provided node. If drain is true, the node is drained before being deleted.
-func (a *Actuator) deleteNode(node *apiv1.Node, drain bool) (result status.NodeDeleteResult) {
-	nodeGroup, err := a.ctx.CloudProvider.NodeGroupForNode(node)
-	if err != nil {
-		return status.NodeDeleteResult{ResultType: status.NodeDeleteErrorInternal, Err: errors.NewAutoscalerError(errors.CloudProviderError, "failed to find node group for %s: %v", node.Name, err)}
-	}
+func (a *Actuator) cleanUpNode(node *apiv1.Node, drain bool) status.NodeDeleteResult {
+	nodeGroup, _ := a.ctx.CloudProvider.NodeGroupForNode(node)
 	if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
 		return status.NodeDeleteResult{ResultType: status.NodeDeleteErrorInternal, Err: errors.NewAutoscalerError(errors.InternalError, "picked node that doesn't belong to a node group: %s", node.Name)}
 	}
-
-	defer func() { a.nodeDeletionTracker.EndDeletion(nodeGroup.Id(), node.Name, result) }()
 	if drain {
-		a.nodeDeletionTracker.StartDeletionWithDrain(nodeGroup.Id(), node.Name)
 		if evictionResults, err := a.evictor.DrainNode(a.ctx, node); err != nil {
 			return status.NodeDeleteResult{ResultType: status.NodeDeleteErrorFailedToEvictPods, Err: err, PodEvictionResults: evictionResults}
 		}
 	} else {
-		a.nodeDeletionTracker.StartDeletion(nodeGroup.Id(), node.Name)
 		if err := a.evictor.EvictDaemonSetPods(a.ctx, node, time.Now()); err != nil {
 			// Evicting DS pods is best-effort, so proceed with the deletion even if there are errors.
 			klog.Warningf("Error while evicting DS pods from an empty node %q: %v", node.Name, err)
 		}
 	}
-
 	if err := WaitForDelayDeletion(node, a.ctx.ListerRegistry.AllNodeLister(), a.ctx.AutoscalingOptions.NodeDeletionDelayTimeout); err != nil {
 		return status.NodeDeleteResult{ResultType: status.NodeDeleteErrorFailedToDelete, Err: err}
 	}
+	return status.NodeDeleteResult{ResultType: status.NodeDeleteInPogress}
+}
 
-	if err := DeleteNodeFromCloudProvider(a.ctx, node, a.clusterState); err != nil {
-		return status.NodeDeleteResult{ResultType: status.NodeDeleteErrorFailedToDelete, Err: err}
+// scheduleDeletion schedule the deletion on of the provided node by adding a node to NodeDeletionBatcher. If drain is true, the node is drained before being deleted.
+func (a *Actuator) scheduleDeletion(node *apiv1.Node, drain bool) {
+	status := a.cleanUpNode(node, drain)
+	if status.Err != nil {
```

Comment:
The exact same error handling repeats in `delete_in_batch.go`, but split into two functions (`process_*`). First off I think this version of having 1 function with `if drain` inside is more readable than having 2 functions. But more importantly, since the 2 places are supposed to do the same thing - can we extract this functionality somehow and call it from both places? Then we're sure the handling actually is the same in both places, and there's less code duplication.

Reply:
Yes, this have similarity since it record events about failed ScaleDown, but scale down failure is different in those cases, so I'm not sure that it worth to create a generalised function. 
Also I don't agree that two functions (process_*) are less readable.

Reply:
Sure, readability is frequently very subjective. However, the two function approach does have more duplication, and worse locality. I know there's not much in these functions anyway, I'm just worried about somebody adding/changing something in the future in one of the 3 places, and forgetting about one of the other ones. I think having all places call 1 function with empty/drain argument (and any other potential differentiators) would make that scenario less likely.

Reply:
Agreed, updated the code.